protobuf: 3.0.0-alpha-3.1 -> 3.0.0-beta-2

eduarrrd opened the pull request:
Also split out gmock's source so that it can be copied into protobuf's
source. Hopefully this hack can be removed again once gmock is
replaced by gtest.

Also introduce my company to the COPYING doc since they own the
copyright on my contributions.

edolstra commented on pkgs/development/libraries/protobuf/3.0.nix:
```diff
@@ -1,12 +1,46 @@
-{ callPackage, fetchFromGitHub, ... } @ args:
+{ stdenv, fetchFromGitHub
+, autoreconfHook, zlib, gmock
+, ...
```
Why `...`?

eduarrrd replied:
Good question. I missed this while merging. Let me remove that.

eduarrrd replied:
Fixed.

pSub commented on COPYING:
```diff
@@ -1,4 +1,4 @@
-Copyright (c) 2003-2006 Eelco Dolstra
+Copyright (c) 2003-2006 Eelco Dolstra, Google Inc.
```
Why Google Inc.?

eduarrrd replied:
Because according my terms of employment say that they own the copyright on anything I produce while I work for them.

If you question is why I'm including them in this file? Because that's part of the terms for me (=them) to contribute here.

What it means? It just naming one (out of many) other copyright holders but Eelco. (Unless there's a CLA that reassigns all copyright to him that I missed.)

jagajaga replied:
I'm totally against that line.

jagajaga replied:
> Because according my terms of employment say that they own the copyright on anything I produce while I work for them.
Can you do that at home at your free time? :)

zimbatm replied:
Copyright on this repo is implicitly shared between all the authors (if your country adheres to the [Berne convention](https://en.wikipedia.org/wiki/Berne_Convention)). It's correct to update the COPYING file to reflect that. Our mistake is in not doing it sooner.

eduarrrd replied:
@jagajaga Thanks for the suggestion but I am doing this at home in my free time. Can you elaborate why the notice is unacceptable to you? Do you believe it would be factually incorrect?

jagajaga replied:
Google Inc. is not related to NixOS.

peti replied:
Listing Google Inc. as the only explicitly mentioned copyright holder next to Eelco does feel excessive. I also doubt that it is necessary to do so. Generally speaking, we need a better policy with regard to (c) issues that what we have today, i.e. mentioned Eelco as the only (c) holder doesn't feel quite right either. Anyway, he's certainly the most important contributor to Nixpkgs, so the status quo feels okay'ish to me until we have figured out a better solution.

eduarrrd replied:
I don't understand this reasoning? In your view, what is this file for? What would have if Eelco quits nixos? Will he be removed from this file?

eduarrrd replied:
@peti I'm not sure why people didn't add their names either. If you want, I can extract the names of all committers from the git log and add them here so there won't be false exclusivity. How does that sound?

peti replied:
The fact that Eelco is the only person mentioned in that file is a historical curiosity. He started the project, he created that file, and he's the only one who has ever touched it since (in 6b0b4586945e3531a04561f4fcffd7e5f666513d). IMHO, the contents of that file is inaccurate and meaningless, at best, and we need to improve the way we handle copyright issues.

Anyway, I agree with @jagajaga that adding Google to that license statement seems odd and I'm not in favor of doing so.

zimbatm replied:
It might be easier to maintain a list of contributors like golang does: https://github.com/golang/go/blob/master/CONTRIBUTORS

kernc replied:
@eduarrrd Google does _not_ hold the copyright to patches you contribute at your home in your free time.

peti replied:
> I can extract the names of all committers from the git log and add them here so there won't be false exclusivity. How does that sound?

I think it would be best to do that (a) in a separate PR and (b) only after discussing this topic on the nix-dev mailing list. This is not something that we can decide easily.

jagajaga replied:
@peti but I think that discussion about exactly this PR is over after @kernc's message?

silverhook replied:
(I’m quite busy, so it’s a hurried reply, sorry)

@eduarrrd If you created that in your free time, it doesn’t count as [“work for hire” (link: Wikipedia)][wfh] even under US copyright.

To quote [Columbia Law School’s article][cls] on this topic:

> On the other hand, if you’ve written a novel or a computer application in your spare time, and that isn’t your job, you, not your employer, own those rights. That’s important to remember for the many creators who also have day jobs.

In any case you would be entitled to moral rights under the Berne convention, but *if* you assigned the rights to the work you did here to Google (e.g. as work for hire or a separate contract), it might indeed be that Google has to be listed as a copyright holder.

Without seeing the details, the situation you describe here, it indeed sounds like you wrote this in your free time and Google has no right to be listed as copyright holder. The one who has the right to be listed is you, who wrote it.

As @zimbatm noticed, all the authors should have been listed in the copyright headers and the `COPYING` or `AUTHORS` file already before. So that needs to be fixed at some point in any case.


[wfh]: https://en.wikipedia.org/wiki/Work_for_hire
[cls]: http://web.law.columbia.edu/keep-your-copyrights/copyrights/works-made-hire

eduarrrd replied:
@kernc I'm not what makes you the expert on my terms of employment but the
top comment might clear things up for you:
https://news.ycombinator.com/item?id=1969979
On Dec 8, 2015 11:38 AM, "kernc" <notifications@github.com> wrote:

> In COPYING
> <https://github.com/NixOS/nixpkgs/pull/11535#discussion_r46977676>:
>
> > @@ -1,4 +1,4 @@
> > -Copyright (c) 2003-2006 Eelco Dolstra
> > +Copyright (c) 2003-2006 Eelco Dolstra, Google Inc.
>
> @eduarrrd <https://github.com/eduarrrd> Google does *not* hold the
> copyright to patches you contribute at your home in your free time.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/11535/files#r46977676>.
>

jagajaga replied:
@eduarrrd send me your code, I will remove that duty from you and commit under my own name :)
Indeed. C'mon. You are person, not company.

peti replied:
@jagajaga,

> I think that discussion about exactly this PR is over after @kernc's message?

I believe @eduarrrd is the only one qualified to judge whether this patch is his private property or that of his employer, so I'd rather not presume to tell him what to do.

What I *can* say is that I, personally, will not merge a patch that modifies the top-level `COPYING` file without getting explicit consent from a representative part of the NixOS community first. If this means that this patch cannot be merged, then that's unfortunate but it's the way it is. Other collaborators might feel differently about this and they might be willing to merge the patch as-is. If that is so, then more power to them! It's just that I won't.

jagajaga replied:
I agree with you wholeheartedly, @peti.

silverhook replied:
@eduarrrd Ah so it’s additional terms of contract. Yes, in US you can probably do that.

It could be argued that you’re not competing with Google with this contribution, but without seeing the contract etc. etc. it’s just guesswork. You’d probably need to get a lawyer, if you wanted to try that.

(also, of course, neither this or my post above is legal advice)

eduarrrd replied:
#11557 is the followup. IF that one gets merged I will remove the COPYING change from this one. Post to nix-dev to follow regarding the #11557 to follow later today.

domenkozar replied:
cc @wkennington who might be interested into this debate.

vcunat replied:
I'm certainly against adding that line.

eduarrrd replied:
@vcunat so is everyone ;-). It currently serves a "do not merge" function.
If #11557 merges this line will not be part of this PR.

On Fri, Dec 11, 2015 at 10:49 AM Vladimír Čunát <notifications@github.com>
wrote:

> In COPYING
> <https://github.com/NixOS/nixpkgs/pull/11535#discussion_r47368887>:
>
> > @@ -1,4 +1,4 @@
> > -Copyright (c) 2003-2006 Eelco Dolstra
> > +Copyright (c) 2003-2006 Eelco Dolstra, Google Inc.
>
> I'm certainly against adding that line.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/11535/files#r47368887>.
>

jagajaga commented on pkgs/development/libraries/protobuf/3.0.nix:
```diff
@@ -1,12 +1,43 @@
+  doCheck = true;
+
+  meta = {
+    description = "Protocol Buffers - Google's data interchange format";
```
Omit pkg's name in the description.

eduarrrd replied:
Thanks, done.

jagajaga commented on pkgs/development/libraries/protobuf/3.0.nix:
```diff
@@ -1,12 +1,43 @@
+         almost all of its internal RPC protocols and file formats.
+      '';
+
+    license = "mBSD";
```
Please use `stdenv.lib.licenses.whateverYouNeed`.

eduarrrd replied:
Thanks, done.

jagajaga commented on pkgs/development/libraries/protobuf/3.0.nix:
```diff
@@ -1,12 +1,43 @@
+      '';
+
+    license = "mBSD";
+
```
Please omit empty new lines in `meta`.

eduarrrd replied:
Thanks, done.

jagajaga commented on pkgs/development/libraries/protobuf/3.0.nix:
```diff
@@ -1,12 +1,43 @@
+
+    license = "mBSD";
+
+    homepage = https://developers.google.com/protocol-buffers/;
```
Please set `platforms   = platforms.unix;`.

eduarrrd replied:
Thanks, done.

jagajaga commented:
Please don't forget to properly name commit following our guides in `CONTRIBUTING.md`/

oconnorr commented:
@eduarrrd you should be able to continue this PR now.

eduarrrd committed:
protobuf: 3.0.0-alpha-3.1 -> 3.0.0-beta-2

Also split out gmock's source so that it can be copied into protobuf's
source. Hopefull this hack can be removed again once gmock is replaced
by gtest.

This does not include python bindings.

eduarrrd commented:
Update for current upstream beta.

PTAL

vcunat commented on pkgs/development/libraries/gmock/default.nix:
```diff
@@ -1,15 +1,13 @@
-
+{ stdenv, cmake, callPackage }:
+let
+  source = callPackage ./source.nix { };
```
Why not use `fetchzip` instead?

vcunat replied:
Or maybe `fetchFromGitHub` directly, as I see the URL redirects to github anyway.

eduarrrd replied:
@vcunat Done (using fetchzip).

Didn't know about fetchzip. Now I do :-). I didn't implement the fetch, I just split out the maintainer's fetching logic into a separate output so I didn't investigate too deeply.

vcunat replied:
That was a partial misunderstanding between us. It's enough to just write
```nix
let
  source = fetchzip { /* ... */ };
```
I don't think there's any benefit in making a `mkDerivation` wrapper around that to only copy the same thing into another path. If you want to override `name`, that can also be passed to `fetchzip` directly.

eduarrrd replied:
I took this pattern from pkgs/development/libraries/gtest/default.nix . Does that one use it better than I do?
Do you have an example of what you are proposing? I'm not sure I understand your proposal correctly.

vcunat replied:
The thing is that `fetchurl` only fetches, but `fetchzip` (and derived functions) also unpacks the archive afterwards, all in a single step. So, `gtest` could also use `fetchzip` instead of the whole gtest-src derivation. It might be just because `fetchzip` is relatively new and anyway, its main purpose is to fight non-deterministic archives.

See https://github.com/eduarrrd/nixpkgs/pull/1

eduarrrd commented:
Adressed @vcunat 's concern.

Tested. Travis fails due to (seemingly) unrelated merge conflicts.

PTAL.

vcunat committed:
gmock: refactor source

eduarrrd commented:
@vcunat Many thanks for your "example" :-).

vcunat commented:
I also enabled parallel building, as I didn't like to wait too long.
